### PR TITLE
Add Metric and supporting classes

### PR DESCRIPTION
**Pull request by c24t**
Like #342, this diff adds a few metrics classes following the [proto spec](https://github.com/census-instrumentation/opencensus-proto/blob/24333298e36590ea0716598caacc8959fc393c48/src/opencensus/proto/metrics/v1/metrics.proto#L33) and [java implementation](https://github.com/census-instrumentation/opencensus-java/tree/08d09c48a40b7dc6b5dba8603795084fc6ea9695/api/src/main/java/io/opencensus/metrics/export). The new classes are `Metric`, `TimeSeries`, and the value class `ValueDistribution`.

This diff is written in the style of #337, following the java package structure. I included preconditions to match the comments in the metrics proto file, but this client's behavior isn't guaranteed to match the java client's.

Addresses #335.

### CODE REVIEW

File: opencensus/metrics/export/metric.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from opencensus.metrics.export.metric_descriptor import MetricDescriptorType
+
+from opencensus.metrics.export.value import ValueLong
+from opencensus.metrics.export.value import ValueDouble
```
Comment (mayurkale22) [inline opencensus/metrics/export/metric.py]: Sort the imports, and import module instead of class.

File: opencensus/metrics/export/metric.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+    :type descriptor: class: '~opencensus.metrics.export.metric_descriptor.MetricDescriptor'  # noqa
+    :param descriptor: The metric's descriptor.
+
+    :type timeseries: list(:class: '~opencensus.metrics.export.time_series.TimeSeries')
+    :param timeseries: One or more timeseries for a single metric, where each
+    timeseries has one or more points.
+    """
+
+    def __init__(self, time_series, descriptor):
```
Comment (mayurkale22) [inline opencensus/metrics/export/metric.py]: nit: s/descriptor/metric_descriptor, also first param should be ```descriptor```.

File: opencensus/metrics/export/metric.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        if not time_series:
+            raise ValueError
```
Comment (mayurkale22) [inline opencensus/metrics/export/metric.py]: error message is missing, something like ```raise ValueError("time_series must not be null")```

File: opencensus/metrics/export/metric.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    :type descriptor: class: '~opencensus.metrics.export.metric_descriptor.MetricDescriptor'  # noqa
+    :param descriptor: The metric's descriptor.
+
+    :type timeseries: list(:class: '~opencensus.metrics.export.time_series.TimeSeries')
+    :param timeseries: One or more timeseries for a single metric, where each
+    timeseries has one or more points.
+    """
+
+    def __init__(self, time_series, descriptor):
+        if not time_series:
```
Comment (mayurkale22) [inline opencensus/metrics/export/metric.py]: Check time_series element not null

File: opencensus/metrics/export/time_series.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    '~opencensus.metrics.label_value.LabelValue')
+    :param label_values: The set of label values that uniquely identify this
+    timeseries.
+
+    :type points: list(:class: '~opencensus.metrics.export.point.Point')
+    :param points: The data points of this timeseries.
+    """
+
+    def __init__(self, start_timestamp, label_values, points):
+        if start_timestamp is None:
```
Comment (mayurkale22) [inline opencensus/metrics/export/time_series.py]: Param should be like label_values, points, start_timestamp

File: opencensus/metrics/export/time_series.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+    This class implements the spec for v1 TimeSeries structs as of
+    opencensus-proto release v0.0.2. See opencensus-proto for details:
+
+        https://github.com/census-instrumentation/opencensus-proto/blob/24333298e36590ea0716598caacc8959fc393c48/src/opencensus/proto/metrics/v1/metrics.proto#L112  # noqa
+
+    A TimeSeries is a collection of data points that describes the time-varying
+    values of a metric.
+
+    :type start_timestamp: int
```
Comment (mayurkale22) [inline opencensus/metrics/export/time_series.py]: type ```time```?

File: opencensus/metrics/export/time_series.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    :param start_timestamp: The time when the cumulative value was reset to
+    zero.
```
Comment (mayurkale22) [inline opencensus/metrics/export/time_series.py]: Please mention ```Must be non-null for cumulative points```

File: opencensus/metrics/export/value.py [hunk 8]
```diff
@@ -86,5 +89,148 @@ class ValueSummary(Value):
...
+
+    Exemplars are example points that may be used to annotate aggregated
+    Distribution values. They are metadata that gives information about a
+    particular value added to a Distribution bucket.
+
+    :type value: double
+    :param value: Value of the exemplar point, determines which bucket the
+    exemplar belongs to.
+
+    :type timestamp: str
```
Comment (mayurkale22) [inline opencensus/metrics/export/value.py]: type ```time```

File: opencensus/metrics/export/value.py [hunk 9]
```diff
@@ -86,5 +89,148 @@ class ValueSummary(Value):
...
+    :param count: The number of values in the population.
+
+    :type sum_: float
+    :param sum_: The sum of the values in the population.
+
+    :type sum_of_squared_deviation: float
+    :param sum_of_squared_deviation: The sum of squared deviations from the
+    mean of the values in the population.
+
+    :type bucket_bounds: list(float)
```
Comment (mayurkale22) [inline opencensus/metrics/export/value.py]: Please add TODO here to change this to ```BucketOptions``` as per new [metric.proto](https://github.com/census-instrumentation/opencensus-proto/blob/master/src/opencensus/proto/metrics/v1/metrics.proto#L204).

File: opencensus/metrics/export/metric.py (hunk 1 above)
Comment (c24t) [reply opencensus/metrics/export/metric.py]: Ah I see the note in the style guide about class imports now, done.

File: opencensus/metrics/export/metric.py (hunk 2 above)
Comment (c24t) [reply opencensus/metrics/export/metric.py]: It's called `descriptor` in the proto def: https://github.com/census-instrumentation/opencensus-proto/blob/24333298e36590ea0716598caacc8959fc393c48/src/opencensus/proto/metrics/v1/metrics.proto#L37.

File: opencensus/metrics/export/metric.py (hunk 4 above)
Comment (c24t) [reply opencensus/metrics/export/metric.py]: This is checking for null and empty series, I'm assuming that's the behavior we want based on ["one or more timeseries for a single metric"](https://github.com/census-instrumentation/opencensus-proto/blob/24333298e36590ea0716598caacc8959fc393c48/src/opencensus/proto/metrics/v1/metrics.proto#L47), but there's no such check in the java client.

File: opencensus/metrics/export/time_series.py (hunk 6 above)
Comment (c24t) [reply opencensus/metrics/export/time_series.py]: What's a `time`? Changed to `str` to match the existing tests for other ts fields, but we should consider using datetimes (or epoch ms/ns) everywhere instead.

File: opencensus/metrics/export/time_series.py (hunk 5 above)
Comment (c24t) [reply opencensus/metrics/export/time_series.py]: How come? I'm following the order in the proto def: https://github.com/census-instrumentation/opencensus-proto/blob/24333298e36590ea0716598caacc8959fc393c48/src/opencensus/proto/metrics/v1/metrics.proto#L112.

File: opencensus/metrics/export/value.py (hunk 8 above)
Comment (c24t) [reply opencensus/metrics/export/value.py]: See above, unless time has some special meaning I'm missing here.

File: opencensus/metrics/export/time_series.py (hunk 5 above)
Comment (mayurkale22) [reply opencensus/metrics/export/time_series.py]: This is to keep things consistent across libraries. Couterapart in Java : https://github.com/census-instrumentation/opencensus-java/blob/master/api/src/main/java/io/opencensus/metrics/export/TimeSeries.java#L53. Also start_timestamp is optional field, but must be non-null for cumulative Points.

Comment (c24t) [reply opencensus/metrics/export/time_series.py]: OK, I changed the field order and added a check in f17f1f1.

File: opencensus/metrics/export/metric.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+        elif self.descriptor.type in (
+                metric_descriptor.MetricDescriptorType.GAUGE_DISTRIBUTION,
+                metric_descriptor.MetricDescriptorType.CUMULATIVE_DISTRIBUTION
+        ):
+            check_type = value.ValueDistribution
+        elif self.descriptor.type == (
+                metric_descriptor.MetricDescriptorType.SUMMARY):
+            check_type = value.ValueSummary
+        else:
+            raise ValueError("Unknown metric descriptor type")
```
Comment (liyanhui1228) [inline opencensus/metrics/export/metric.py]:
```
# This relation mapping could be put at the top of this file
DESCRIPTOR_VALUE = {
    metric_descriptor.MetricDescriptorType.GAUGE_INT64: value.ValueLong,
    metric_descriptor.MetricDescriptorType.CUMULATIVE_INT64: value.ValueLong,
    ...
}

check_type = DESCRIPTOR_VALUE.get(self.descriptor.type)
if check_type is None:
    raise ValueError("Unknown metric descriptor type")
```

**Pull request merged**